File: app/src/main/java/io/maido/m8client/AudioDevice.java
package io.maido.m8client;

import android.media.AudioDeviceInfo;

import androidx.annotation.NonNull;

import java.util.Objects;

public class AudioDevice {

    private final String name;
    private final int type;
    private final int deviceId;

    public AudioDevice(AudioDeviceInfo info) {
        this.name = info.getProductName().toString();
        this.type = info.getType();
        this.deviceId = info.getId();
    }

    @NonNull
    @Override
    public String toString() {
        return name + " " + type;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AudioDevice that = (AudioDevice) o;
        return type == that.type && deviceId == that.deviceId && Objects.equals(name, that.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, type, deviceId);
    }

    public int getDeviceId() {
        return deviceId;
    }
}

File: app/src/main/kotlin/io/maido/m8client/AudioDevice.kt
package io.maido.m8client

import android.media.AudioDeviceInfo
import java.util.*

data class AudioDevice(
    private val name: String,
    private val type: Int,
    val deviceId: Int,
) {
    constructor(info: AudioDeviceInfo) : this(info.productName.toString(), info.type, info.id)

    override fun toString() = "$name $type"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AudioDevice
        return type == that.type && deviceId == that.deviceId && name == that.name
    }

    override fun hashCode() = Objects.hash(name, type, deviceId)
}